File: core/async-rpc-retrying-caller.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include "core/async-rpc-retrying-caller.h"

#include <folly/Conv.h>
#include <folly/ExceptionWrapper.h>
#include <folly/Format.h>
#include <folly/Logging.h>
#include <folly/Unit.h>

#include "connection/rpc-client.h"
#include "core/async-connection.h"
#include "core/hbase-rpc-controller.h"
#include "core/region-location.h"
#include "core/result.h"
#include "exceptions/exception.h"
#include "if/HBase.pb.h"
#include "utils/connection-util.h"
#include "utils/sys-util.h"
#include "utils/time-util.h"

using folly::exception_wrapper;

namespace hbase {

template <typename RESP>
AsyncSingleRequestRpcRetryingCaller<RESP>::AsyncSingleRequestRpcRetryingCaller(
    std::shared_ptr<AsyncConnection> conn, std::shared_ptr<folly::HHWheelTimer> retry_timer,
    std::shared_ptr<hbase::pb::TableName> table_name, const std::string& row,
    RegionLocateType locate_type, Callable<RESP> callable, std::chrono::nanoseconds pause,
    uint32_t max_retries, std::chrono::nanoseconds operation_timeout_nanos,
    std::chrono::nanoseconds rpc_timeout_nanos, uint32_t start_log_errors_count)
    : conn_(conn),
      retry_timer_(retry_timer),
      table_name_(table_name),
      row_(row),
      locate_type_(locate_type),
      callable_(callable),
      pause_(pause),
      max_retries_(max_retries),
      operation_timeout_nanos_(operation_timeout_nanos),
      rpc_timeout_nanos_(rpc_timeout_nanos),
      start_log_errors_count_(start_log_errors_count),
      promise_(std::make_shared<folly::Promise<RESP>>()),
      tries_(1) {
  controller_ = conn_->CreateRpcController();
  start_ns_ = TimeUtil::GetNowNanos();
  max_attempts_ = ConnectionUtils::Retries2Attempts(max_retries);
  exceptions_ = std::make_shared<std::vector<ThrowableWithExtraContext>>();
}

template <typename RESP>
AsyncSingleRequestRpcRetryingCaller<RESP>::~AsyncSingleRequestRpcRetryingCaller() {}

template <typename RESP>
folly::Future<RESP> AsyncSingleRequestRpcRetryingCaller<RESP>::Call() {
  auto f = promise_->getFuture();
  LocateThenCall();
  return f;
}

template <typename RESP>
void AsyncSingleRequestRpcRetryingCaller<RESP>::LocateThenCall() {
  int64_t locate_timeout_ns;
  if (operation_timeout_nanos_.count() > 0) {
    locate_timeout_ns = RemainingTimeNs();
    if (locate_timeout_ns <= 0) {
      CompleteExceptionally();
      return;
    }
  } else {
    locate_timeout_ns = -1L;
  }

  conn_->region_locator()
      ->LocateRegion(*table_name_, row_, locate_type_, locate_timeout_ns)
      .then([this](std::shared_ptr<RegionLocation> loc) { Call(*loc); })
      .onError([this](const exception_wrapper& e) {
        OnError(e,
                [this, e]() -> std::string {
                  return "Locate '" + row_ + "' in " + table_name_->namespace_() + "::" +
                         table_name_->qualifier() + " failed with e.what()=" +
                         e.what().toStdString() + ", tries = " + std::to_string(tries_) +
                         ", maxAttempts = " + std::to_string(max_attempts_) + ", timeout = " +
                         TimeUtil::ToMillisStr(operation_timeout_nanos_) + " ms, time elapsed = " +
                         TimeUtil::ElapsedMillisStr(this->start_ns_) + " ms";
                },
                [](const exception_wrapper& error) {});
      });
}

template <typename RESP>
void AsyncSingleRequestRpcRetryingCaller<RESP>::OnError(
    const exception_wrapper& error, Supplier<std::string> err_msg,
    Consumer<exception_wrapper> update_cached_location) {
  ThrowableWithExtraContext twec(error, TimeUtil::GetNowNanos());
  exceptions_->push_back(twec);
  if (!ExceptionUtil::ShouldRetry(error) || tries_ >= max_retries_) {
    CompleteExceptionally();
    return;
  }

  if (tries_ > start_log_errors_count_) {
    LOG(WARNING) << err_msg();
  } else {
    VLOG(1) << err_msg();
  }

  int64_t delay_ns;
  if (operation_timeout_nanos_.count() > 0) {
    int64_t max_delay_ns = RemainingTimeNs() - ConnectionUtils::kSleepDeltaNs;
    if (max_delay_ns <= 0) {
      CompleteExceptionally();
      return;
    }
    delay_ns = std::min(max_delay_ns, ConnectionUtils::GetPauseTime(pause_.count(), tries_ - 1));
  } else {
    delay_ns = ConnectionUtils::GetPauseTime(pause_.count(), tries_ - 1);
  }
  update_cached_location(error);
  tries_++;

  /*
   * The HHWheelTimer::scheduleTimeout() fails with an assertion from
   * EventBase::isInEventBaseThread() if we execute the schedule in a random thread, or one of
   * the IOThreadPool threads (with num threads > 1). I think there is a bug there in using retry
   * timer from IOThreadPool threads. It only works when executed from a single-thread pool
   * (retry_executor() is). However, the scheduled "work" which is the LocateThenCall() should
   * still happen in a thread pool, that is why we are submitting the work to the CPUThreadPool.
   * IOThreadPool cannot be used without fixing the blocking call that we do at TCP connection
   * establishment time (see ConnectionFactory::Connect()), otherwise, the IOThreadPool thread
   * just hangs because it deadlocks itself.
   */
  conn_->retry_executor()->add([&]() {
    retry_timer_->scheduleTimeoutFn(
        [this]() { conn_->cpu_executor()->add([&]() { LocateThenCall(); }); },
        std::chrono::milliseconds(TimeUtil::ToMillis(delay_ns)));
  });
}

template <typename RESP>
void AsyncSingleRequestRpcRetryingCaller<RESP>::Call(const RegionLocation& loc) {
  int64_t call_timeout_ns;
  if (operation_timeout_nanos_.count() > 0) {
    call_timeout_ns = this->RemainingTimeNs();
    if (call_timeout_ns <= 0) {
      this->CompleteExceptionally();
      return;
    }
    call_timeout_ns = std::min(call_timeout_ns, rpc_timeout_nanos_.count());
  } else {
    call_timeout_ns = rpc_timeout_nanos_.count();
  }

  std::shared_ptr<RpcClient> rpc_client;

  rpc_client = conn_->rpc_client();

  ResetController(controller_, call_timeout_ns);

  // TODO: RegionLocation should propagate through these method chains as a shared_ptr.
  // Otherwise, it may get deleted underneat us. We are just copying for now.
  auto loc_ptr = std::make_shared<RegionLocation>(loc);
  callable_(controller_, loc_ptr, rpc_client)
      .then([loc_ptr, this](const RESP& resp) { this->promise_->setValue(std::move(resp)); })
      .onError([&, loc_ptr, this](const exception_wrapper& e) {
        OnError(
            e,
            [&, this, e]() -> std::string {
              return "Call to " + folly::sformat("{0}:{1}", loc_ptr->server_name().host_name(),
                                                 loc_ptr->server_name().port()) +
                     " for '" + row_ + "' in " + loc_ptr->DebugString() + " of " +
                     table_name_->namespace_() + "::" + table_name_->qualifier() +
                     " failed with e.what()=" + e.what().toStdString() + ", tries = " +
                     std::to_string(tries_) + ", maxAttempts = " + std::to_string(max_attempts_) +
                     ", timeout = " + TimeUtil::ToMillisStr(this->operation_timeout_nanos_) +
                     " ms, time elapsed = " + TimeUtil::ElapsedMillisStr(this->start_ns_) + " ms";
            },
            [&, this](const exception_wrapper& error) {
              conn_->region_locator()->UpdateCachedLocation(*loc_ptr, error);
            });
      });
}

template <typename RESP>
void AsyncSingleRequestRpcRetryingCaller<RESP>::CompleteExceptionally() {
  this->promise_->setException(RetriesExhaustedException(tries_ - 1, exceptions_));
}

template <typename RESP>
int64_t AsyncSingleRequestRpcRetryingCaller<RESP>::RemainingTimeNs() {
  return operation_timeout_nanos_.count() - (TimeUtil::GetNowNanos() - start_ns_);
}

template <typename RESP>
void AsyncSingleRequestRpcRetryingCaller<RESP>::ResetController(
    std::shared_ptr<HBaseRpcController> controller, const int64_t& timeout_ns) {
  controller->Reset();
  if (timeout_ns >= 0) {
    controller->set_call_timeout(std::chrono::milliseconds(
        std::min(static_cast<int64_t>(INT_MAX), TimeUtil::ToMillis(timeout_ns))));
  }
}

// explicit instantiations for the linker. Otherwise, you have to #include the .cc file for the
// templetized
// class definitions.
class OpenScannerResponse;
template class AsyncSingleRequestRpcRetryingCaller<std::shared_ptr<hbase::Result>>;
template class AsyncSingleRequestRpcRetryingCaller<folly::Unit>;
template class AsyncSingleRequestRpcRetryingCaller<std::shared_ptr<OpenScannerResponse>>;
template class AsyncSingleRequestRpcRetryingCaller<bool>;
} /* namespace hbase */

File: hbase-client/src/main/java/org/apache/hadoop/hbase/client/AsyncRpcRetryingCaller.java
/**
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.hadoop.hbase.client;

import static org.apache.hadoop.hbase.client.ConnectionUtils.SLEEP_DELTA_NS;
import static org.apache.hadoop.hbase.client.ConnectionUtils.getPauseTime;
import static org.apache.hadoop.hbase.client.ConnectionUtils.resetController;
import static org.apache.hadoop.hbase.client.ConnectionUtils.translateException;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;
import java.util.function.Supplier;
import org.apache.hadoop.hbase.CallQueueTooBigException;
import org.apache.hadoop.hbase.DoNotRetryIOException;
import org.apache.hadoop.hbase.NotServingRegionException;
import org.apache.hadoop.hbase.TableName;
import org.apache.hadoop.hbase.TableNotEnabledException;
import org.apache.hadoop.hbase.TableNotFoundException;
import org.apache.hadoop.hbase.exceptions.ScannerResetException;
import org.apache.hadoop.hbase.ipc.HBaseRpcController;
import org.apache.hadoop.hbase.util.EnvironmentEdgeManager;
import org.apache.hadoop.hbase.util.FutureUtils;
import org.apache.yetus.audience.InterfaceAudience;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.apache.hbase.thirdparty.io.netty.util.Timer;

@InterfaceAudience.Private
public abstract class AsyncRpcRetryingCaller<T> {

  private static final Logger LOG = LoggerFactory.getLogger(AsyncRpcRetryingCaller.class);

  private final Timer retryTimer;

  private final int priority;

  private final long startNs;

  private final long pauseNs;

  private final long pauseForCQTBENs;

  private int tries = 1;

  private final int maxAttempts;

  private final int startLogErrorsCnt;

  private final List<RetriesExhaustedException.ThrowableWithExtraContext> exceptions;

  private final long rpcTimeoutNs;

  protected final long operationTimeoutNs;

  protected final AsyncConnectionImpl conn;

  protected final CompletableFuture<T> future;

  protected final HBaseRpcController controller;

  public AsyncRpcRetryingCaller(Timer retryTimer, AsyncConnectionImpl conn, int priority,
      long pauseNs, long pauseForCQTBENs, int maxAttempts, long operationTimeoutNs,
      long rpcTimeoutNs, int startLogErrorsCnt) {
    this.retryTimer = retryTimer;
    this.conn = conn;
    this.priority = priority;
    this.pauseNs = pauseNs;
    this.pauseForCQTBENs = pauseForCQTBENs;
    this.maxAttempts = maxAttempts;
    this.operationTimeoutNs = operationTimeoutNs;
    this.rpcTimeoutNs = rpcTimeoutNs;
    this.startLogErrorsCnt = startLogErrorsCnt;
    this.future = new CompletableFuture<>();
    this.controller = conn.rpcControllerFactory.newController();
    this.controller.setPriority(priority);
    this.exceptions = new ArrayList<>();
    this.startNs = System.nanoTime();
  }

  private long elapsedMs() {
    return TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs);
  }

  protected final long remainingTimeNs() {
    return operationTimeoutNs - (System.nanoTime() - startNs);
  }

  protected final void completeExceptionally() {
    future.completeExceptionally(new RetriesExhaustedException(tries - 1, exceptions));
  }

  protected final void resetCallTimeout() {
    long callTimeoutNs;
    if (operationTimeoutNs > 0) {
      callTimeoutNs = remainingTimeNs();
      if (callTimeoutNs <= 0) {
        completeExceptionally();
        return;
      }
      callTimeoutNs = Math.min(callTimeoutNs, rpcTimeoutNs);
    } else {
      callTimeoutNs = rpcTimeoutNs;
    }
    resetController(controller, callTimeoutNs, priority);
  }

  private void tryScheduleRetry(Throwable error) {
    long pauseNsToUse = error instanceof CallQueueTooBigException ? pauseForCQTBENs : pauseNs;
    long delayNs;
    if (operationTimeoutNs > 0) {
      long maxDelayNs = remainingTimeNs() - SLEEP_DELTA_NS;
      if (maxDelayNs <= 0) {
        completeExceptionally();
        return;
      }
      delayNs = Math.min(maxDelayNs, getPauseTime(pauseNsToUse, tries - 1));
    } else {
      delayNs = getPauseTime(pauseNsToUse, tries - 1);
    }
    tries++;
    retryTimer.newTimeout(t -> doCall(), delayNs, TimeUnit.NANOSECONDS);
  }

  protected Optional<TableName> getTableName() {
    return Optional.empty();
  }

  protected final void onError(Throwable t, Supplier<String> errMsg,
      Consumer<Throwable> updateCachedLocation) {
    if (future.isDone()) {
      // Give up if the future is already done, this is possible if user has already canceled the
      // future. And for timeline consistent read, we will also cancel some requests if we have
      // already get one of the responses.
      LOG.debug("The future is already done, canceled={}, give up retrying", future.isCancelled());
      return;
    }
    Throwable error = translateException(t);
    // We use this retrying caller to open a scanner, as it is idempotent, but we may throw
    // ScannerResetException, which is a DoNotRetryIOException when opening a scanner as now we will
    // also fetch data when opening a scanner. The intention here is that if we hit a
    // ScannerResetException when scanning then we should try to open a new scanner, instead of
    // retrying on the old one, so it is declared as a DoNotRetryIOException. But here we are
    // exactly trying to open a new scanner, so we should retry on ScannerResetException.
    if (error instanceof DoNotRetryIOException && !(error instanceof ScannerResetException)) {
      future.completeExceptionally(error);
      return;
    }
    if (tries > startLogErrorsCnt) {
      LOG.warn(errMsg.get() + ", tries = " + tries + ", maxAttempts = " + maxAttempts +
        ", timeout = " + TimeUnit.NANOSECONDS.toMillis(operationTimeoutNs) +
        " ms, time elapsed = " + elapsedMs() + " ms", error);
    }
    updateCachedLocation.accept(error);
    RetriesExhaustedException.ThrowableWithExtraContext qt =
      new RetriesExhaustedException.ThrowableWithExtraContext(error,
        EnvironmentEdgeManager.currentTime(), "");
    exceptions.add(qt);
    if (tries >= maxAttempts) {
      completeExceptionally();
      return;
    }
    // check whether the table has been disabled, notice that the check will introduce a request to
    // meta, so here we only check for disabled for some specific exception types.
    if (error instanceof NotServingRegionException || error instanceof RegionOfflineException) {
      Optional<TableName> tableName = getTableName();
      if (tableName.isPresent()) {
        FutureUtils.addListener(conn.getAdmin().isTableDisabled(tableName.get()), (disabled, e) -> {
          if (e != null) {
            if (e instanceof TableNotFoundException) {
              future.completeExceptionally(e);
            } else {
              // failed to test whether the table is disabled, not a big deal, continue retrying
              tryScheduleRetry(error);
            }
            return;
          }
          if (disabled) {
            future.completeExceptionally(new TableNotEnabledException(tableName.get()));
          } else {
            tryScheduleRetry(error);
          }
        });
      } else {
        tryScheduleRetry(error);
      }
    } else {
      tryScheduleRetry(error);
    }
  }

  protected abstract void doCall();

  CompletableFuture<T> call() {
    doCall();
    return future;
  }
}
